### ISSUE TITLE
`default` for `TypeVar` is not working consistently with type inference for generic class

### ISSUE THREAD

**Issue by kc0506** [microsoft/pyright#9158]
<h3> Environment data </h3>

-   Language Server version: 2024.9.2
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python
from dataclasses import dataclass, field
from typing import Generic, Literal, assert_type, overload, reveal_type

from typing_extensions import TypeVar

FlagA = TypeVar("FlagA", Literal[True], Literal[False], default=Literal[False])


@dataclass
class Config(Generic[FlagA]):
    flag_a: FlagA = field(default=False)


class HasFeatureA(Config[FlagA], Generic[FlagA]):
    pass


T = TypeVar("T", Config[Literal[True]], Config[Literal[False]], contravariant=True)


@dataclass
class A(Generic[T]):
    config: T

    @overload
    def f(self: "A[HasFeatureA[Literal[True]]]") -> Literal[1]: ...
    @overload
    def f(self: "A[HasFeatureA[Literal[False]]]") -> Literal[2]: ...

    def f(self) -> int: ...


a1 = A(Config(flag_a=True))
assert_type(a1.f(), Literal[1])

a2 = A(config=Config(flag_a=False))
assert_type(a2.f(), Literal[2])

default_config = Config()
a3 = A(config=default_config)
assert_type(a3.f(), Literal[2])

a4 = A(config=Config())
assert_type(a4.f(), Literal[2])  # "assert_type" mismatch: expected "Literal[2]" but received "Literal[1]"
reveal_type(a4)  # A[Config[Literal[True]]]
```

<h3> Expected behavior </h3>

All four calls to `assert_type` should not raise an error. Specifically, `a3` and `a4` should both be of type `A[Config[Literal[False]]]`.

<h3> Actual behavior </h3>

`a4` is inferred as `A[Config[Literal[True]]]`, causing `assert_type` to fail.

#### Remove constraints of `T`
This way all the `assert_type` works, but we lose the type hint for `A.config`.

#### Use `bound` instead of constraints on `T`
Does not work because bounds cannot be generic, which means the upper bound of `T` becomes `Config[Literal[False]]`.

Please let me know if there are other working workarounds, thanks.

**Comment by kc0506** [comment]
Thank you for the explanation; I now understand why the code doesn't work. However, I think there's another issue: the type-checker doesn’t recognize the `__init__` signatures of `dataclass` the same way they appear at runtime.

For example, consider this code:
```py
class A[T: bool]:
    def __init__(self, a: T = False):
        self.a = a

a: A[Literal[True]] = A()  # correctly raises an error

@dataclass
class B[T: bool]:
    a: T = field(default=False)

b: B[Literal[True]] = B()  # does not raise an error

print(signature(B))  # (a: T = False) -> None
```

At runtime, the signatures of `A.__init__` and `B.__init__` are the same, but the type-checker handles them differently. For `A()`, the default value in `__init__` allows it to be inferred as `A[Literal[False]]`, which correctly raises an error when assigned to `A[Literal[True]]`. However, in the case of `B()`, the type-checker infers it as `B[Unknown]` instead of `B[Literal[False]]`, and it doesn’t raise an error.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.384
